### PR TITLE
Add electromagnetic storm missions

### PR DESCRIPTION
**Pull request by Jacunn**
The Ember Waste is a unique region of space, but it's mostly just a different colour and has lots of special wormholes. What if there were missions that took place here, that involve electromagnetic storms? Essentially, the concept is that overtime, charged particles in the waste reach higher concentrations in certain regions, resulting in electromagnetic storms. It adds a new twist to deal with in a combat mission that I've included in the Ember Waste. This is probably far from ideal in it's current state, but it's a proof of concept.

### CODE REVIEW

**Comment by Amazinite** [comment]
Related: https://github.com/endless-sky/endless-sky/issues/1553, https://github.com/endless-sky/endless-sky/issues/1552#issuecomment-246232974

File: emstorm [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	variant
+		"EMStorm" 22
+		
+# Mission
+mission "Remnant: Stranded Korath"
+	landing
+	source
+		government "Remnant"
+	name "Korath ships stranded in Remnant Territory"
+	description "There are Korath fleets located in the Caeculus, Stercutus and Segesta systems. Find and eliminate them, then return to <planet> for payment."
```
Comment (tehhowch) [inline emstorm]: You can use #3336 here.

File: emstorm [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		has "event: remnant: void sprite research"
+		has "Remnant: Key Stones: done"
+		has "Remnant: Defense 3: done"
+		has "Remnant: Bounty: done" >= 3
+	waypoint Caeculus
+	waypoint Stercutus
+	waypoint Segesta
+	on offer
+		conversation
+			`As you land, you are greeted by a man and a woman, both dressed in quite a striking manner, perhaps as a sign of distinction. They collectively begin to chant in an almost gleeful tone`
```
Comment (tehhowch) [inline emstorm]: You are missing a **lot** of punctuation in your conversation text, particularly before character speech.

File: emstorm [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+					accept
+				`	"Sorry, but that sounds too dangerous."`
+					decline
+	npc evade
+		personality
+			heroic
+			vindictive
+		government "Korath Nanobots"
+		system Caeculus
+		fleet "EMStorm"
```
Comment (tehhowch) [inline emstorm]: A `fleet` with this name is never defined.

**Comment by Zitchas** [comment]
I like the idea of having electromagnetic storms in the Ember Wastes. It'd be neat to have something other than the background colour to change up the environmental factors.

**Comment by Jacunn** [comment]
Oops, I refactored names poorly! I'll fix this up and resubmit it! :)

**Comment by tehhowch** [comment]
Don't resubmit it, just add commits to the branch you are using that address the issues.

If you resubmit the PR, you separate the new one from all previous feedback (and also pollute the repo, since you can't delete an issue or PR, just close it).

**Comment by robot-rover** [comment]
Perhaps similar to the way the swarm of nanobots from the Korath work?

File: emstorm [hunk 4]
```diff
@@ -45,29 +42,29 @@ outfit "Storm"
...
+fleet "Electromagnetic (Moderate)"
 	government "Korath Nanobots"
 	personality
 		heroic
 		vindictive
 	variant
-		"EMStorm" 15
+		"Electromagnetic Storm" 15
 		
 fleet "Electromagnetic(strong)"
```
Comment (KiLEdEnNis) [inline emstorm]: you made
> fleet "Electromagnetic (Strong)"

so put a space and capitalize the s in "strong" to make that work

**Comment by Kriegstofu** [comment]
keep in mind that this might influence/interfere the gascraft mission

**Comment by Jacunn** [comment]
As it stands the storms will not kick in until after the void sprite mission has been completed, among others. This is to ensure that there won't be any interference with other missions that are possibly future story-line critical.

**Comment by Zitchas** [comment]
I'd like to dig into the content a bit more to see how this is actually working, but my thoughts are so far:
I really like the idea of having some kind of "natural hazard" in the Ember Wastes. Ideally, something that is fairly nasty but not necessarily lethal throughout the Ember Wastes, but weaker (or possibly almost non-existant) in the core Remnant systems. Which would make the Remnant both protected by it and penned in by it (and thus is becomes fodder for missions of them trying to figure out a way to reduce its effect on their ships). That being said, I'd prefer to have it in effect right from the beginning of the game - it would help explain why Hai explorers or adventurous humans with access to the crystals haven't bothered to spend more time in the Ember Wastes. After all, the Remnant got through that wormhole. If they can, someone else should have been able to too. That being said, having it start out at a low level of danger and ramp up at some point would be even better, as that provides for a different kind of challenge and mission set than the usual empire vs empire conflict.

As far as the gascraft and other missions go, those can be adapted or perhaps modified somehow. Maybe some way to give the gascraft a form of immunity to it?

**Comment by Amazinite** [comment]
I imagine an EM storm should only do shield damage, not hull damage, so the Gascraft should be effected much given that you’re not surviving more than one shot from the Archon anyway.

**Comment by Zitchas** [comment]
Yeah, or I was thinking of a combo of shield damage and ionization.

**Comment by Jacunn** [comment]
I agree that the changes and considerations you all mentioned are
necessary. I more-so created this as a proof of concept for this idea, and
I'd be happier for someone with more experience and/or knowledge to take
this over as long as I'm given a mention for the concept. Ideally I'd like
this to be a system condition rather than a fleet to give space in general
a livelier feeling. Some ideas I have in light of what you have all
mentioned, however:
-As per the gas craft concerns, once this is implemented as a recurring
phenomenon, a flag can be set during this mission to prevent the storms
from occurring.
-As per the obscene levels of damage, I agree that the effects of these
storms must be lowered, otherwise they can just become crippling and
boring, rather than interesting and a new "mechanic".
-For graphical purposes, I do not have the talent to add any visual effects
towards the storm. If anyone would like to do this, that would be amazing.
-For a storm "dissipating" after a while, perhaps the storm has limited
energy capacity in the first place and slowly recharges, meaning that a
storm will have long recharge periods.
What do you all think? If nobody wishes to take this on, I will endeavour
to update this when I have the time do so.

<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by Zitchas** [comment]
I'm interested in pursuing this, as it ties in nicely with my Remnant stuff. I think I'd rather keep it separate from my current PR as I think that's pretty much ready to merge, but this is something of a separate idea that will take some fine tuning before it is ready.

That being said, I have been testing some different damage values, and I've got some additional ideas for it, including at least one idea for making it a more "system-wide" type effect, even if I don't have the coding knowledge to make it into an actual game mechanic.

**Comment by endless-sky** [comment]
It seems to me that random damage / disruption with no visible cause would just be annoying. At least with the nanobots there's dialog that mentions what's gnawing away at your ship.

If there's going to be environmental hazards, they should be visible in various ways. For example, an EM storm could cause occasional flashes of lightning in the "background" of the system, or it could cause your radar to flicker in and out.

**Comment by Zitchas** [comment]
That would definitely be a great environmental effect. Part of the reason I'd like to see it attempted in the Waste is that it already has the ominous red colour, so I feel that players wouldn't be as surprised to have something happen. If we could get some kind of general system type damage, it would almost certainly be merged with a short mission along the same lines as the nanobot one. I would be happy to contribute that part if someone else can do the code for having a continuous damage effect.

I agree having lightning flashes or something else in the background would be nice, but between the omnious red background and a dialogue advisory, I don't think it is actually necessary.

**Comment by Amazinite** [comment]
Closing this and moving the discussion on system/environmental hazards as a whole to #4116.

**Pull request closed without merging**